File: functions/device/update_device.py
from datetime import datetime, timezone

from firebase_admin import firestore
from flask import Request
from models.constants import Collections, DeviceFields
from models.data_models import Device
from utils.logging_utils import get_logger


def update_device(request: Request):
    """
    Update the device ID for the authenticated user.

    Args:
        request: The Flask request object with user_id attached from authentication.

    Returns:
        A Device object containing the updated device information.
    """
    logger = get_logger(__name__)
    logger.info(f"Updating device for user {request.user_id}")

    # Get validated data from request
    device_data_input = request.validated_params

    current_time = datetime.now(timezone.utc)

    # Reference to the user's device document
    db = firestore.client()
    device_ref = db.collection(Collections.DEVICES).document(request.user_id)

    # Create or update the device document
    device_data = {
        DeviceFields.DEVICE_ID: device_data_input.device_id,
        DeviceFields.UPDATED_AT: current_time,
    }

    # Update the device document
    device_ref.set(device_data, merge=True)
    logger.info(f"Device updated for user {request.user_id}")

    # Create and return a Device object
    return Device(
        device_id=device_data_input.device_id, updated_at=current_time.isoformat() + "Z"
    )

File: functions/src/device/update-device.ts
import { Request, Response } from 'express';
import { DocumentData, getFirestore, Timestamp, UpdateData, } from 'firebase-admin/firestore';
import { Collections, DeviceFields } from '../models/constants.js';
import { Device, DevicePayload } from '../models/data-models.js';
import { getLogger } from '../utils/logging-utils.js';
import { formatTimestamp } from '../utils/timestamp-utils.js';

import { fileURLToPath } from 'url';
import path from 'path';

const __filename = fileURLToPath(import.meta.url);
const logger = getLogger(path.basename(__filename));

/**
 * Update the device ID for the authenticated user.
 *
 * @param req - The Express request object containing:
 *              - userId: The authenticated user's ID (attached by authentication middleware)
 *              - validated_params: The validated request parameters containing:
 *                - device_id: The new device ID to set
 * @param res - The Express response object
 *
 * @returns A Device object containing the updated device information
 */
export const updateDevice = async (
  req: Request,
  res: Response,
): Promise<void> => {
  const currentUserId = req.userId;
  logger.info(`Updating device for user ${currentUserId}`);

  // Get validated data from request
  const deviceDataInput = req.validated_params as DevicePayload;

  const currentTime = Timestamp.now();

  // Reference to the user's device document
  const db = getFirestore();
  const deviceRef = db.collection(Collections.DEVICES).doc(currentUserId);

  // Create or update the device document
  const deviceData: UpdateData<DocumentData> = {
    [DeviceFields.DEVICE_ID]: deviceDataInput.device_id,
    [DeviceFields.UPDATED_AT]: currentTime,
  };

  // Update the device document
  await deviceRef.set(deviceData, { merge: true });
  logger.info(`Device updated for user ${currentUserId}`);

  // Create and return a Device object
  const device: Device = {
    device_id: deviceDataInput.device_id,
    updated_at: formatTimestamp(currentTime),
  };

  res.json(device);
};
